tracing.go

```
@@ -84,53 +87,155 @@ func (s Span) ParentID() uint64 { return s.parentID }
+	if spanCtx != nil {
+		span := c.Tracer().StartSpan("", ext.RPCServerOption(spanCtx))
+		ext.PeerService.Set(span, callReq.Headers[CallerName])
+		span.SetTag("as", callReq.Headers[ArgScheme])
+		setPeerHostPort(span, c.remotePeerInfo.HostPort)
+		return span
+	}
+	if err != opentracing.ErrUnsupportedFormat && err != opentracing.ErrSpanContextNotFound {
+		c.log.Error("Failed to extract Zipkin-style span: " + err.Error())
 	}
+	return nil
 }
 
-func newSpan(traceID, spanID, parentID uint64, tracingEnabled bool) *Span {
-	flags := byte(0)
-	if tracingEnabled {
-		flags = tracingFlagEnabled
+// ExtractInboundSpan is a higher level version of extractInboundSpan().
+// If the lower-level attempt to create a span from incoming request was
+// successful (e.g. when then Tracer supports Zipkin-style trace IDs),
+// then the application headers are only used to read the Baggage and add
+// it to the existing span. Otherwise, the standard OpenTracing API supported
+// by all tracers is used to deserialize the tracing context from the
+// application headers and start a new server-side span.
+// Once the span is started, it is wrapped in a new Context, which is returned.
+func ExtractInboundSpan(ctx context.Context, call *InboundCall, headers map[string]string, tracer opentracing.Tracer) context.Context {
+	var span = call.Response().span
+	if span != nil {
+		if headers != nil {
+			// extract SpanContext from headers, but do not start another span with it,
+			// just get the baggage and copy to the already created span
+			carrier := opentracing.TextMapCarrier(headers)
+			if sc, _ := tracer.Extract(opentracing.TextMap, carrier); sc != nil {
+				sc.ForeachBaggageItem(func(k, v string) bool {
+					span.Context().SetBaggageItem(k, v)
+					return true
+				})
+			}
+		}
+	} else {
+		var parent opentracing.SpanContext
+		if headers != nil {
+			carrier := opentracing.TextMapCarrier(headers)
+			parent, _ = tracer.Extract(opentracing.TextMap, carrier)
```

errors being ignored here?